draft-ietf-tls-esni.md

```diff
@@ -446,33 +490,114 @@ to harmonize these to make sure that we maintain key separation.]]
+verify the certificate with the public name, as follows:
+
+- If the server resumed a session or negotiated parameters not based on
+  certificates, the client MUST abort the connection with an illegal_parameter
+  alert. This case is invalid because {{client-behavior}} requires the client
+  to only offer ESNI-established sessions, and {{server-behavior}} requires
+  the server to decline ESNI-established sessions if it did not accept ESNI.
+
+- The client MUST verify that the certificate is valid for ESNIKeys.public_name.
+  If invalid, it MUST abort the connection with the appropriate alert.
+
+- If the server requests a client certificate, the client MUST respond with an
+  empty Certificate message, denoting no client certificate.
+
+Note that verifying a connection for the public name does not verify it for the
+origin. The TLS implementation MUST NOT surface such connections as successful to
+the application. It additionally MUST ignore all session tickets and session IDs
+presented by the server. These connections are only used to trigger retries, as
+described in {{client-behavior}}. This may be implemented, for instance, by
+reporting a failed connection with a dedicated error code.
+
+## Client-Facing Server Behavior {#server-behavior}
 
 Upon receiving an "encrypted_server_name" extension, the client-facing
-server MUST first perform the following checks:
-
-- If it is unable to negotiate TLS 1.3 or greater, it MUST
-  abort the connection with a "handshake_failure" alert.
-
-- If the ClientEncryptedSNI.record_digest value does not match the cryptographic
-  hash of any known ESNIKeys structure, it MUST abort the connection with
-  an "illegal_parameter" alert. This is necessary to prevent downgrade attacks.
-  [[OPEN ISSUE: We looked at ignoring the extension but concluded
-  this was better.]]
+server MUST check that it is able to negotiate TLS 1.3 or greater. If not,
+it MUST abort the connection with a "handshake_failure" alert.
+
+If the ClientEncryptedSNI.record_digest value does not match the
+cryptographic hash of any known ESNIKeys structure, it MUST ignore the
+extension and proceed with the connection, with the following added
+behavior:
+
+- It MUST include the "encrypted_server_name" extension in
+  EncryptedExtensions message with the "response_type" field set to
+  "esni_retry_requested" and the "retry_keys" field set to one or more
+  ESNIKeys structures with up-to-date keys. Servers MAY supply multiple
+  ESNIKeys values of different versions. This allows a server to support
+  multiple versions at once.
+
+- The server MUST ignore all PSK identities in the ClientHello which correspond
+  to ESNI PSKs. ESNI PSKs offered by the client are associated with the ESNI
+  name. The server was unable to decrypt then ESNI name, so it should not resume
+  them when using the cleartext SNI name. This restriction allows a client to
+  reject resumptions in {{verify-public-name}}.
```

Comment:
Why not prohibit resumption and PSK entirely for the case where the keys need replacing?  This only works for certificate-based server authentication anyway.  Are you really trying to salvage this for servers that authenticate with a PSK?